opengever/maintenance/scripts/use_canonical_document_ids.py

```diff
@@ -0,0 +1,135 @@
+import os
+import transaction
+import urllib
+
+
+SEPARATOR = '-' * 78
+
+
+REPORT_PATH = os.path.abspath(
+    os.path.join(__file__,
+                 '..', '..', '..', '..',
+                 'renamed_documents.csv'))
+
+
+def dev_break_document_ids(options):
+    """DEV-method, use to break the document ids.
+
+    """
+    catalog = api.portal.get_tool('portal_catalog')
+    brains = catalog.unrestrictedSearchResults(
+        portal_type='opengever.document.document')
+
+    for number, brain in enumerate(brains):
+        old_name = brain.id
+        new_name = "with spaces {}".format(number)
+        obj = brain.getObject()
+        api.content.rename(obj=obj, new_id=new_name)
+        print "broke {} to {} ({})".format(old_name, new_name,
+                                           '/'.join(obj.getPhysicalPath()))
+
+    if not options.dry_run:
+        transaction.commit()
+    print "done breaking."
+
+
+def use_canonical_document_id(options):
+    """Rename documents that do not use the current naming-schema.
+
+    Also write a short report of the renamed documents to a csv file.
+
+    """
+    catalog = api.portal.get_tool('portal_catalog')
+    brains = catalog.unrestrictedSearchResults(
+        portal_type='opengever.document.document')
+
+    to_rename = []
+
+    for brain in brains:
+        name = brain.id
+        if name.startswith('template-') or name.startswith('document-'):
+            if options.verbose:
+                print "skipped: {}".format(brain.getPath())
+            continue
+
+        to_rename.append(brain.getObject())
+
+    if to_rename:
+        print SEPARATOR
+
+    def build_url(fragment):
+        fragment = urllib.quote(fragment)
+        if not options.domain:
+            return fragment
+
+        # we assume a virtualhost and strip the plone site id
+        assert fragment.startswith("/"), "path should start with /"
+        visible_fragment = "/".join(fragment.split('/')[2:])
+        base = "https://{}/".format(options.domain)
+        return urljoin(base, visible_fragment)
+
+    with open(REPORT_PATH, 'w+') as csvfile:
+        writer = DictWriter(csvfile, fieldnames=["old_url", "new_url"])
+        writer.writeheader()
+
+        # elevated privileges are necessary to be able to modify stuff in
+        # closed dossiers.
+        with elevated_privileges():
+            for obj in to_rename:
+                old_name = obj.getId()
+                old_url = '/'.join(obj.getPhysicalPath())
+                name_chooser = IDocumentNameFromTitle(obj)
+                # returns unicode, this is considered invalid by plone :-o
+                new_name = str(name_chooser.title)
+                print "renaming '{}' to '{}' ({})".format(
+                    old_name, new_name, '/'.join(obj.getPhysicalPath()))
+                api.content.rename(obj=obj, new_id=new_name)
+                new_url = '/'.join(obj.getPhysicalPath())
+                writer.writerow({
```

Comment:
Instead of just writing a report - could you please also log the mutations? Then they will be included in `instance0.log`, timestamped and archived in a well defined place.